### PR TITLE
Decouple cart presenter into lazy array

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR decouples the cart presentation process into dedicated getters, implementing the lazy array mechanism. This is a first step for improvement performance in FO (although its effect is not visible yet, because all getters are called at the prestashop javascript object creation, I will deal with this later, it's a big piece)
| Type?             | improvement
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | CI green
| UI Tests          | https://github.com/matthieu-rolland/ga.tests.ui.pr/actions/runs/9713325661
| Fixed issue or discussion?     | 
| Related PRs       | 
| Sponsor company   | PrestashopSA

### CODE REVIEW

**Review by boherm — DISMISSED** [review]

File: src/Adapter/Presenter/Cart/CartLazyArray.php
Comment (Hlavtox) [inline src/Adapter/Presenter/Cart/CartLazyArray.php]: You need to specificaly check for a null or !isset.

If the cart is empty, the condition will pass and execute the query again.

Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartLazyArray.php]: oops, my bad 👍 fixed, I'm so used to using `!empty` in other cases...  

(just in case, all getters are executed only once here, because the AbstractLazyArray caches getters's result, but yes indeed it would execute the count query when it's not needed 👍  )

File: src/Adapter/Presenter/Cart/CartLazyArray.php
Comment (matks) [inline src/Adapter/Presenter/Cart/CartLazyArray.php:27]: License header is missing

File: src/Adapter/Presenter/Cart/CartPresenter.php
Comment (matks) [inline src/Adapter/Presenter/Cart/CartPresenter.php]: This is a BC break, please document it

**Review by matks — DISMISSED** [review]

File: src/Adapter/Presenter/Cart/CartPresenter.php
Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartPresenter.php]: Eventually I moved back those functions into CartPresenter, because it was breaking, I also had to set them as public, because they now must be called from the lazy cart...

File: controllers/front/CartController.php
Comment (matthieu-rolland) [inline controllers/front/CartController.php:143]: I had to remove this, because setting a value on the lazy array makes a fatal error

Also since this is a lazy array, data is fetched on demand, no need to preset the data.

This is called when adding a product to Cart in FO, I tested and it works. @jolelievre do you think I should instantiate a new CartPresenter here, in case products would have already been set and would consequently have outdated data ?

Comment (matthieu-rolland) [reply controllers/front/CartController.php:143]: answered, we'll add a class extending AbstractLazyArray, allowing to set data, only for the CartPresenter for now

we'll be done in a next PR

**Review by boherm — DISMISSED** [review]

File: controllers/front/CartController.php
Comment (Hlavtox) [reply controllers/front/CartController.php:143]: @matthieu-rolland If you make it extend AbstractLazyArray, you can just do

```
$presentedCart->offsetUnset('key', true);
$presentedCart->__set('key', $dataToAssign);
```

But I am not really sure what this piece of code should do. What does it filter? 🤔

Comment (matthieu-rolland) [reply controllers/front/CartController.php:143]: yes we just saw that...  but today modules can update this "array" directly, we want to avoid a breaking change on this, especially because this is passed to a hook

on this PR right now it works only if there was no previously set data, otherwise it triggers a fatal error, this is a BC that could be annoying

(the filter's purpose is to send back only required data I guess)

We just need to set `$force` to true in the `__set` method, only for cart (at least for now, I'm not sure if it could be interesting for other presenters...)

Comment (Hlavtox) [reply controllers/front/CartController.php:143]: @matthieu-rolland 

> yes we just saw that... but today modules can update this "array" directly, we want to avoid a breaking change on this, especially because this is passed to a hook
> 
> on this PR right now it works only if there was no previously changed data, otherwise it triggers a fatal error, this is a BC that could be annoying

What if we modify the __set method on the AbstractLazyArray to make this work? Probably just unset before the set? That would remove the BC break and make working with it easier.

> (the filter's purpose is to send back only required data I guess)

Yeah, it filters the array and keeps only the keys in a list - https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Filter/FrontEndObject/ProductFilter.php, not sure how needed/useful this is.

Comment (matthieu-rolland) [reply controllers/front/CartController.php:143]: (this will be done in the next Pull Request)

File: src/Adapter/Presenter/Cart/CartLazyArray.php
Comment (Hlavtox) [inline src/Adapter/Presenter/Cart/CartLazyArray.php:236]: We could cache it like this on repeated requests

```suggestion
        if ($this->productsCount === null) {
            // If product list is already available, no need to execute a new sql query
            if (isset($this->products)) {
                $this->productsCount = array_reduce($this->products, function ($count, $product) {
                    return $count + $product['quantity'];
                }, 0);
            // Product list is not available, only query the nb of products
            } else {
                $this->productsCount = (int) $this->cart->getNbProducts($this->cart->id);
            }
        }
```

File: src/Adapter/Presenter/Cart/CartLazyArray.php
Comment (Hlavtox) [inline src/Adapter/Presenter/Cart/CartLazyArray.php:317]: Wrapping this in `if ($this->discounts === null)` will avoid repeating the logic on another access.

File: controllers/front/CartController.php
Comment (matthieu-rolland) [reply controllers/front/CartController.php:143]: I want to make another class that extends `AbstractLazyArray` to modify the `__set` method (wether we unset data or pass the `$force` paramater at true in it does the same thing I think ?)

So that we can allow overriding values only on the presenters we want, using the abstract class or its extension (by default you can't override and it's a wanted behavior I think)

File: src/Adapter/Presenter/Cart/CartLazyArray.php
Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartLazyArray.php:317]: then I have to declare $discounts like this: `private array|null $discounts = null;`, which is a bit annoying

otherwise I get an error of the kind `must not be accessed before initialization` 

but in this case it's interesting indeed 👍

Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartLazyArray.php:317]: ah but I forgot, there's no need, the results of all these getters are cached in AbstractLazyArray, it's the point (among other things)

I mean except if these getter are called directly, but it's not how it's supposed to be, those presenters are meant to be used as arrays in templates

File: src/Adapter/Presenter/Cart/CartLazyArray.php
Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartLazyArray.php:236]: same as below: these getters are not to be called directly, the presenter is meant to be used as an array, in templates, or having its properties called directly

the AbstractLazyArray is responsible for calling the getters, only if the data is not cached already

File: src/Adapter/Presenter/Cart/CartLazyArray.php
Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartLazyArray.php:317]: maybe all these getters should have protected accesses instead of public 🤔

Comment (Hlavtox) [reply src/Adapter/Presenter/Cart/CartLazyArray.php:317]: I didn't know this at all... :D

So, when I have been working on product presenters, checks like this are useless? https://github.com/PrestaShop/PrestaShop/blob/a9e361099940601237ba1f554a9f2fba76488cf9/src/Adapter/Presenter/Product/ProductLazyArray.php#L354

Comment (matthieu-rolland) [reply src/Adapter/Presenter/Cart/CartLazyArray.php:317]: yep :D  if it used as an array, as it should be

but not if you call the getter directly, but we shouldn't call the getter directly, only `AbstractLazyArray` should call them...  that's why I wonder wether we should put those getters as protected instead...

File: controllers/front/CartController.php
Comment (PrestaEdit) [reply controllers/front/CartController.php:143]: If it's a class using the LazyArray feature, you can't simply use this ?

`$presentedCart->offsetSet('products', $this->get('prestashop.core.filter.front_end_object.product_collection')
                ->filter($presentedCart['products']))`

Setting the third params as `true` will replace the possible existing methods `getProducts()`on this LazyArray by it's simple value.

Also, there is a method called `presentProducts` inside the `CartPresenter`; it's params is an array of products rows. Maybe you can call it with the filtered products ?
(It's only a suggestion, I don't check really more about it, I hope it's not a false way to go :))

Comment (matthieu-rolland) [reply controllers/front/CartController.php:143]: hello !

well it's a matter of bc break, the cart presenter used to be an array, so people could update its value simply by setting the values on the array.

Now that it's a lazy array, it's an object that behaves like an array, except you can't rewrite already set values, which is a BC break, especially when this object is passed to the CartPresenter's hook

so I'd want it to work just like before for devs, without having to call `offsetSet` directly

File: src/Adapter/Presenter/AbstractLazyArray.php
Comment (matthieu-rolland) [inline src/Adapter/Presenter/AbstractLazyArray.php:71]: this is temporary, a next PR will replace all usages of phpdoc in this class with new php attribute feature, there will be no need for regex

File: controllers/front/CartController.php
Comment (jolelievre) [inline controllers/front/CartController.php:138]: I thought we discussed about the possibility to set an optional product filtered here, so that the API correctly returns only whitelisted fields of products (only in this use case)

**Review by boherm — APPROVED** [review]

File: controllers/front/CartController.php
Comment (matthieu-rolland) [reply controllers/front/CartController.php:138]: I'll do it in the next pull request, that will implement php attributes and add an attribute to allow overriding value in the array manner

but if you want I can, just for the time between this PR and the next one, put the intermediary class that allows editing, I did it earlier but I removed the commit

**Review by jolelievre — APPROVED** [review]

**Comment by ps-jarvis** [comment]
PR merged, well done!

Message to @PrestaShop/committers: do not forget to milestone it before the merge.

**Pull request merged**
